File: packages/typescript/sdk/consumer/src/tools.ts
import type { SlopNode } from "./types";

export interface LlmTool {
  type: "function";
  function: {
    name: string;
    description: string;
    parameters: Record<string, unknown>;
  };
}

export interface ChatMessage {
  role: "system" | "user" | "assistant" | "tool";
  content: string;
  tool_calls?: { id: string; type: "function"; function: { name: string; arguments: string } }[];
  tool_call_id?: string;
}

export interface ToolSet {
  tools: LlmTool[];
  /**
   * Resolve a tool name back to invoke coordinates.
   *
   * For singleton tools: returns `{ path: "/some/path", action: "delete" }`
   * For grouped tools: returns `{ path: null, action: "delete", targets: ["/a", "/b", ...] }`
   *   — caller must extract `target` from args and use it as the path.
   */
  resolve(toolName: string): { path: string | null; action: string; targets?: string[] } | null;
}

/**
 * Walk the state tree and collect all affordances as LLM tools.
 *
 * Affordances that share the same `action` name AND identical `params` schema
 * are **grouped** into a single tool with a `target` parameter. This avoids
 * registering N tools when N nodes share the same action (e.g. 500 cards each
 * with `edit`). Instead, one `edit` tool is registered and the caller supplies
 * the node path via `target`.
 *
 * **Why group by action + schema (not node type or label)?**
 * SLOP's `invoke(path, action, params)` already includes the path — the
 * provider dispatches internally based on path + action. From the consumer
 * side, `delete` IS one operation; the `target` path differentiates
 * "delete card" from "delete comment". The LLM has the state tree in context
 * and can pick the correct path.
 *
 * Singleton affordances (unique action on a single node) keep the original
 * `{nodeId}__{action}` naming with a fixed path.
 *
 * Returns a `ToolSet` with the tools and a `resolve()` function that maps
 * tool names back to invoke coordinates.
 */
export function affordancesToTools(node: SlopNode, path = ""): ToolSet {
  // Phase 1: collect all affordances with their raw data
  const entries: AffordanceEntry[] = [];
  collectAffordances(node, path, [], entries);

  // Phase 2: group by action + canonical param schema
  const groups = groupByActionAndSchema(entries);

  // Phase 3: build tools + resolve map
  const resolveMap = new Map<string, { path: string | null; action: string; targets?: string[] }>();
  const tools: LlmTool[] = [];

  // Detect action-name collisions across groups (same action, different schemas)
  const actionNameCounts = new Map<string, number>();
  for (const group of groups) {
    const action = sanitize(group[0].action);
    actionNameCounts.set(action, (actionNameCounts.get(action) ?? 0) + 1);
  }

  // Track which action names we've used for disambiguation
  const actionNameUsed = new Map<string, number>();

  for (const group of groups) {
    const first = group[0];
    const safeAction = sanitize(first.action);

    if (group.length === 1) {
      // Singleton — keep original naming: {nodeId}__{action}
      const safeId = sanitize(first.nodeId);
      const toolName = `${safeId}__${safeAction}`;
      resolveMap.set(toolName, { path: first.path || "/", action: first.action });
      tools.push({
        type: "function",
        function: {
          name: toolName,
          description: buildDescription(first),
          parameters: first.aff.params ? first.aff.params : { type: "object", properties: {} },
        },
      });
    } else {
      // Grouped — one tool with `target` parameter
      let toolName = safeAction;

      // Disambiguate if multiple groups share the same action name (different schemas)
      if ((actionNameCounts.get(safeAction) ?? 0) > 1) {
        const idx = actionNameUsed.get(safeAction) ?? 0;
        actionNameUsed.set(safeAction, idx + 1);
        if (idx > 0) {
          // Use first entry's nodeId for disambiguation
          toolName = `${safeAction}__${sanitize(first.nodeId)}`;
        }
      }

      const targets = group.map((e) => e.path || "/");
      const baseParams = first.aff.params
        ? JSON.parse(JSON.stringify(first.aff.params))
        : { type: "object", properties: {} };

      // Add `target` as a required parameter
      if (!baseParams.properties) baseParams.properties = {};
      baseParams.properties.target = {
        type: "string",
        description:
          `Path to the target node (e.g. ${targets[0]}).` +
          ` See the state tree for valid paths.`,
      };
      if (!baseParams.required) baseParams.required = [];
      baseParams.required = ["target", ...baseParams.required];

      const isDangerous = group.some((e) => e.aff.dangerous);

      resolveMap.set(toolName, { path: null, action: first.action, targets });
      tools.push({
        type: "function",
        function: {
          name: toolName,
          description: buildGroupDescription(group, isDangerous),
          parameters: baseParams,
        },
      });
    }
  }

  return {
    tools,
    resolve(toolName: string) {
      return resolveMap.get(toolName) ?? null;
    },
  };
}

// ---------------------------------------------------------------------------
// Internal types
// ---------------------------------------------------------------------------

interface AffordanceEntry {
  nodeId: string;
  nodeType: string;
  path: string;
  action: string;
  ancestors: string[];
  aff: any;
  schemaKey: string;
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

/** Sanitize an ID segment for use in tool names (alphanumeric + underscore only). */
function sanitize(s: string): string {
  return s.replace(/[^a-zA-Z0-9]/g, "_");
}

/** Build a canonical key for a param schema (order-independent). */
function canonicalSchemaKey(params: any): string {
  if (!params) return "";
  return JSON.stringify(sortKeysDeep(params));
}

/** Recursively sort object keys for deterministic stringification. */
function sortKeysDeep(obj: any): any {
  if (obj === null || typeof obj !== "object") return obj;
  if (Array.isArray(obj)) return obj.map(sortKeysDeep);
  const sorted: Record<string, any> = {};
  for (const key of Object.keys(obj).sort()) {
    sorted[key] = sortKeysDeep(obj[key]);
  }
  return sorted;
}

/** Recursively collect all affordances from the tree. */
function collectAffordances(
  node: SlopNode,
  path: string,
  ancestors: string[],
  out: AffordanceEntry[],
): void {
  for (const aff of node.affordances ?? []) {
    out.push({
      nodeId: node.id,
      nodeType: node.type,
      path,
      action: aff.action,
      ancestors: ancestors.map(sanitize),
      aff,
      schemaKey: canonicalSchemaKey(aff.params),
    });
  }
  for (const child of node.children ?? []) {
    collectAffordances(child, `${path}/${child.id}`, [...ancestors, node.id], out);
  }
}

/**
 * Group affordance entries by action + param schema.
 *
 * Entries with identical `action` name AND identical canonical param schema
 * are placed in the same group. Different schemas produce separate groups,
 * even if the action name is the same (e.g. `edit` on cards vs `edit` on
 * comments with different param shapes).
 */
function groupByActionAndSchema(entries: AffordanceEntry[]): AffordanceEntry[][] {
  const groups = new Map<string, AffordanceEntry[]>();
  for (const entry of entries) {
    const key = `${entry.action}\0${entry.schemaKey}`;
    const list = groups.get(key) ?? [];
    list.push(entry);
    groups.set(key, list);
  }
  return Array.from(groups.values());
}

/** Build description for a singleton tool. */
function buildDescription(entry: AffordanceEntry): string {
  let desc =
    `${entry.aff.label ?? entry.aff.action}` +
    `${entry.aff.description ? ": " + entry.aff.description : ""}` +
    ` (on ${entry.path || "/"})`;
  if (entry.aff.dangerous) desc += " [DANGEROUS - confirm first]";
  return desc;
}

/** Build description for a grouped tool. */
function buildGroupDescription(group: AffordanceEntry[], isDangerous: boolean): string {
  const first = group[0];
  let desc = first.aff.label ?? first.aff.action;
  if (first.aff.description) desc += `: ${first.aff.description}`;
  desc += ` (${group.length} targets)`;
  if (isDangerous) desc += " [DANGEROUS - confirm first]";
  return desc;
}


/** Format the state tree as readable string for LLM context */
export function formatTree(node: SlopNode, indent = 0): string {
  const pad = "  ".repeat(indent);
  const props = node.properties ?? {};
  const meta = node.meta ?? {};
  const displayName = (props.label ?? props.title) as string | undefined;
  // Always show the node ID. If there's a human-readable label/title, show both.
  const header = displayName && displayName !== node.id
    ? `${node.id}: ${displayName}`
    : node.id;
  const extra = Object.entries(props)
    .filter(([k]) => k !== "label" && k !== "title")
    .map(([k, v]) => `${k}=${JSON.stringify(v)}`)
    .join(", ");
  const affordances = (node.affordances ?? [])
    .map(a => {
      let s = a.action;
      if (a.params?.properties) {
        const params = Object.entries(a.params.properties as Record<string, any>)
          .map(([k, v]) => `${k}: ${v.type}`).join(", ");
        s += `(${params})`;
      }
      return s;
    }).join(", ");

  let line = `${pad}[${node.type}] ${header}`;
  if (extra) line += ` (${extra})`;

  // Show summary from meta (for stubs, lazy nodes, windowed collections)
  if (meta.summary) line += `  — "${meta.summary}"`;

  // Show salience when present
  const salience = meta.salience as number | undefined;
  if (salience != null) line += `  salience=${Math.round(salience * 100) / 100}`;

  if (affordances) line += `  actions: {${affordances}}`;

  const lines = [line];

  // Show window/lazy indicators
  const childCount = node.children?.length ?? 0;
  const totalChildren = meta.total_children as number | undefined;
  if (totalChildren != null && totalChildren > childCount) {
    const window = meta.window as [number, number] | undefined;
    if (window) {
      lines.push(`${pad}  (showing ${childCount} of ${totalChildren})`);
    } else if (childCount === 0) {
      lines.push(`${pad}  (${totalChildren} ${totalChildren === 1 ? "child" : "children"} not loaded)`);
    }
  }

  for (const child of node.children ?? []) {
    lines.push(formatTree(child, indent + 1));
  }
  return lines.join("\n");
}

File: packages/swift/slop-ai/Sources/SlopAI/Tools.swift
import Foundation

public struct LLMTool: Equatable {
  public struct Function: Equatable {
    public var name: String
    public var description: String
    public var parameters: JSONSchema

    public init(name: String, description: String, parameters: JSONSchema) {
      self.name = name
      self.description = description
      self.parameters = parameters
    }
  }

  public var type: String
  public var function: Function

  public init(type: String = "function", function: Function) {
    self.type = type
    self.function = function
  }
}

public struct ToolResolution: Equatable {
  public var path: String?
  public var action: String
  public var targets: [String]?

  public init(path: String?, action: String, targets: [String]? = nil) {
    self.path = path
    self.action = action
    self.targets = targets
  }
}

public struct ToolSet {
  public var tools: [LLMTool]
  private var resolveMap: [String: ToolResolution]

  public init(tools: [LLMTool], resolveMap: [String: ToolResolution]) {
    self.tools = tools
    self.resolveMap = resolveMap
  }

  public func resolve(_ toolName: String) -> ToolResolution? {
    resolveMap[toolName]
  }
}

private struct AffordanceEntry {
  var nodeID: String
  var nodeType: String
  var path: String
  var action: String
  var affordance: Affordance
  var schemaKey: String
}

public func affordancesToTools(_ node: SlopNode, path: String = "") -> ToolSet {
  var entries: [AffordanceEntry] = []
  collectAffordances(node, path: path, entries: &entries)

  let groups = Dictionary(grouping: entries) { "\($0.action)\u{0}\($0.schemaKey)" }
    .values
    .map { Array($0) }

  var actionNameCounts: [String: Int] = [:]
  for group in groups {
    guard let first = group.first else { continue }
    let action = sanitize(first.action)
    actionNameCounts[action, default: 0] += 1
  }

  var actionNameUsed: [String: Int] = [:]
  var resolveMap: [String: ToolResolution] = [:]
  var tools: [LLMTool] = []
  var usedToolNames: Set<String> = []

  for group in groups.sorted(by: {
    let lhs = $0.first.map { "\($0.action)\u{0}\($0.schemaKey)\u{0}\($0.nodeID)" } ?? ""
    let rhs = $1.first.map { "\($0.action)\u{0}\($0.schemaKey)\u{0}\($0.nodeID)" } ?? ""
    return lhs < rhs
  }) {
    guard let first = group.first else { continue }
    let safeAction = sanitize(first.action)

    if group.count == 1 {
      let toolName = reserveToolName("\(sanitize(first.nodeID))__\(safeAction)", used: &usedToolNames)
      resolveMap[toolName] = ToolResolution(path: first.path.isEmpty ? "/" : first.path, action: first.action)
      tools.append(
        LLMTool(
          function: .init(
            name: toolName,
            description: buildDescription(first),
            parameters: first.affordance.params ?? JSONSchema(type: "object", properties: [:])
          )
        )
      )
      continue
    }

    var toolName = safeAction
    if (actionNameCounts[safeAction] ?? 0) > 1 {
      let used = actionNameUsed[safeAction] ?? 0
      actionNameUsed[safeAction] = used + 1
      if used > 0 {
        toolName = "\(safeAction)__\(sanitize(first.nodeID))"
      }
    }
    toolName = reserveToolName(toolName, used: &usedToolNames)

    let targets = group.map { $0.path.isEmpty ? "/" : $0.path }
    let baseParams = cloneSchema(first.affordance.params ?? JSONSchema(type: "object", properties: [:]))
    var properties = baseParams.properties ?? [:]
    properties["target"] = JSONSchema(
      type: "string",
      description: "Path to the target node (e.g. \(targets[0])). See the state tree for valid paths."
    )
    baseParams.properties = properties
    baseParams.required = ["target"] + (baseParams.required ?? [])

    resolveMap[toolName] = ToolResolution(path: nil, action: first.action, targets: targets)
    tools.append(
      LLMTool(
        function: .init(
          name: toolName,
          description: buildGroupDescription(group),
          parameters: baseParams
        )
      )
    )
  }

  return ToolSet(tools: tools, resolveMap: resolveMap)
}

public func formatTree(_ node: SlopNode, indent: Int = 0) -> String {
  let pad = String(repeating: "  ", count: indent)
  let properties = node.properties ?? [:]
  let displayName = properties["label"]?.stringValue ?? properties["title"]?.stringValue
  let header = displayName != nil && displayName != node.id ? "\(node.id): \(displayName!)" : node.id
  let extra = properties.keys
    .sorted()
    .filter { $0 != "label" && $0 != "title" }
    .map { "\($0)=\(canonicalJSON(properties[$0]!))" }
    .joined(separator: ", ")

  let affordances = (node.affordances ?? [])
    .map { affordance in
      var string = affordance.action
      if let properties = affordance.params?.properties {
        let params = properties.keys.sorted().map { "\($0): \(properties[$0]!.type)" }.joined(separator: ", ")
        string += "(\(params))"
      }
      return string
    }
    .joined(separator: ", ")

  var line = "\(pad)[\(node.type)] \(header)"
  if !extra.isEmpty {
    line += " (\(extra))"
  }
  if let summary = node.meta?.summary {
    line += "  — \"\(summary)\""
  }
  if let salience = node.meta?.salience {
    line += "  salience=\((salience * 100).rounded() / 100)"
  }
  if !affordances.isEmpty {
    line += "  actions: {\(affordances)}"
  }

  var lines = [line]
  let childCount = node.children?.count ?? 0
  if let totalChildren = node.meta?.totalChildren, totalChildren > childCount {
    if node.meta?.window != nil {
      lines.append("\(pad)  (showing \(childCount) of \(totalChildren))")
    } else if childCount == 0 {
      lines.append("\(pad)  (\(totalChildren) \(totalChildren == 1 ? "child" : "children") not loaded)")
    }
  }
  for child in node.children ?? [] {
    lines.append(formatTree(child, indent: indent + 1))
  }
  return lines.joined(separator: "\n")
}

private func collectAffordances(_ node: SlopNode, path: String, entries: inout [AffordanceEntry]) {
  for affordance in node.affordances ?? [] {
    entries.append(
      AffordanceEntry(
        nodeID: node.id,
        nodeType: node.type,
        path: path,
        action: affordance.action,
        affordance: affordance,
        schemaKey: canonicalSchemaKey(affordance.params)
      )
    )
  }
  for child in node.children ?? [] {
    collectAffordances(child, path: "\(path)/\(child.id)", entries: &entries)
  }
}

private func sanitize(_ value: String) -> String {
  String(value.unicodeScalars.map { scalar -> Character in
    let value = scalar.value
    let isASCIIAlphaNumeric = (48...57).contains(value) || (65...90).contains(value) || (97...122).contains(value)
    return isASCIIAlphaNumeric ? Character(String(scalar)) : "_"
  })
}

private func reserveToolName(_ base: String, used: inout Set<String>) -> String {
  if used.insert(base).inserted {
    return base
  }
  var suffix = 2
  while !used.insert("\(base)__\(suffix)").inserted {
    suffix += 1
  }
  return "\(base)__\(suffix)"
}

private func canonicalSchemaKey(_ schema: JSONSchema?) -> String {
  guard let schema else { return "" }
  return canonicalJSON(wireJSON(schema))
}

private func buildDescription(_ entry: AffordanceEntry) -> String {
  var description = "\(entry.affordance.label ?? entry.affordance.action)"
  if let detail = entry.affordance.description {
    description += ": \(detail)"
  }
  description += " (on \(entry.path.isEmpty ? "/" : entry.path))"
  if entry.affordance.dangerous == true {
    description += " [DANGEROUS - confirm first]"
  }
  return description
}

private func buildGroupDescription(_ group: [AffordanceEntry]) -> String {
  guard let first = group.first else { return "" }
  var description = first.affordance.label ?? first.affordance.action
  if let detail = first.affordance.description {
    description += ": \(detail)"
  }
  description += " (\(group.count) targets)"
  if group.contains(where: { $0.affordance.dangerous == true }) {
    description += " [DANGEROUS - confirm first]"
  }
  return description
}

private func cloneSchema(_ schema: JSONSchema) -> JSONSchema {
  (try? decodeJSONValue(wireJSON(schema), as: JSONSchema.self)) ?? JSONSchema(type: schema.type)
}
